### ISSUE TITLE
Hausdorff Distance / Loss function

### ISSUE THREAD

**Issue by innat** [keras-team/keras-cv#330]
From wiki

> In mathematics, the Hausdorff distance, or Hausdorff metric, also called Pompeiu–Hausdorff distance, **measures how far two subsets of a metric space are from each other**. It turns the set of non-empty compact subsets of a metric space into a metric space in its own right.

It can be used for point cloud similarity metrics, keypoint detection, segmentation task, etc. For example, in an ongoing kaggle comp [**UW-Madison GI Tract Image Segmentation**](https://www.kaggle.com/competitions/uw-madison-gi-tract-image-segmentation/overview/evaluation), a 3D version of Hausdorff distance is used as one of the metrics to evaluate. 

> Hausdorff distance is a method for calculating the distance between segmentation objects A and B, by calculating the furthest point on object A from the nearest point on object B.

I think it's a useful loss function to have. Here are some references implementation:

- in [tf.keras](https://stackoverflow.com/questions/61897779/how-to-use-hausdorff-metric-in-keras)
- in [pytorch](https://github.com/javiribera/locating-objects-without-bboxes/blob/e51f75ec925f67bb4054ab4a6e9b20d1d8e689cc/object-locator/losses.py#L120)

**Comment by blackhat-coder** [comment]
Hi @innat, I'm interested in implementing this

**Comment by innat** [comment]
@qlzh727 @LukeWood WDYT?

**Comment by bhack** [comment]
It is not in the Keras loss API format but we could be inspired also by other TF ecosystem official implementations:

https://github.com/tensorflow/graphics/blob/master/tensorflow_graphics/nn/loss/hausdorff_distance.py

**Comment by blackhat-coder** [comment]
Hi @innat
Currently there's no `Losses` Folder in `keras-cv` to house loss functions
Do I have to create that ? 

[113](https://github.com/keras-team/keras-cv/issues/113) @LukeWood stated to implement IoU Loss to set an example for loss functions

**Comment by innat** [comment]
Yes, I think so. You need to create a loss folder.

**Comment by blackhat-coder** [comment]
Hi, sorry I haven't made a PR on this issue, I've worked on it but currently waiting for [Dice loss PR](https://github.com/keras-team/keras-cv/pull/371) to get merged to avoid conflicts from me creating `losses`.
I'm also writing test cases (noob ✌).

**Comment by innat** [comment]
Don't worry about that. You can safely do your own. No conflict will happen on this.

**Comment by LukeWood** [comment]
The KerasCV team is attempting to narrow the package scope in order to more efficiently use our limited resources.  As such, we are only accepting contributions directly related to items on the 6 month roadmap for the package.

Feel free to read more on the roadmap:
https://github.com/keras-team/keras-cv/blob/master/.github/ROADMAP.md

**Comment by bhack** [comment]
> The KerasCV team is attempting to narrow the package scope in order to more efficiently use our limited resources. As such, we are only accepting contributions directly related to items on the 6 month roadmap for the package.
> 
> Feel free to read more on the roadmap: https://github.com/keras-team/keras-cv/blob/master/.github/ROADMAP.md

I think that we need to check related PR when we close tickets https://github.com/keras-team/keras-cv/pull/414
